Refactor Error Handling

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Removes the legacy types, interfaces and logic surrounding errors (`types/errors.go`) -- i.e. paying down technical debt.

**To Reviewer**:

I cannot stress how critical this PR is to the safety of the SDK. To aid in review, the work is split into two PRs.

1. This PR handles the critical and core changes to `/types` and `/baseapp`. Reviewers should primarily review changes to `BaseApp` logic, mainly `runTx` and `runMsgs`. The two critical changes (credit @ethanfrey):
  a. `runTx`: We now return `GasInfo, *Result, error`, where `GasInfo` is _always_ returned. The happy path is a non-nil `Result` where tx execution succeeded and an `error` is returned for the sad path.
  b. `runMsgs`: We return now a `*Result, error`. The happy path is a non-nil `Result` where message execution succeeded and where an `error` is returned for the sad path.

2. A subsequent [PR](https://github.com/cosmos/cosmos-sdk/pull/5429) handles updates to the modules and clients.


closes: #4844
replaces: #5072

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

alexanderbez commented:
@ethanfrey this is a revival of the original PR (I did not feel like going through and fixing all the conflicts). I've incorporated your feedback from the original PR for `runTx` and `runMsgs`. Do you mind taking another look at `baseapp.go` when you can spare some time?

fedekunze commented on baseapp/abci.go:
```diff
@@ -368,7 +415,7 @@ func handleQueryStore(app *BaseApp, path []string, req abci.RequestQuery) abci.R
 }
 
-func handleQueryP2P(app *BaseApp, path []string, _ abci.RequestQuery) (res abci.ResponseQuery) {
+func handleQueryP2P(app *BaseApp, path []string, _ abci.RequestQuery) abci.ResponseQuery {
```
We can remove the unused param
```suggestion
func handleQueryP2P(app *BaseApp, path []string) abci.ResponseQuery {
```

fedekunze commented on types/errors/errors.go:
```diff
@@ -268,6 +271,15 @@ func WithType(err error, obj interface{}) error {
 	return Wrap(err, fmt.Sprintf("%T", obj))
 }
 
+func QueryResult(err error) abci.ResponseQuery {
```
nit, godoc

fedekunze commented on types/decimal.go:
```diff
@@ -33,6 +34,13 @@ var (
 
+// Decimal errors
+var (
+	ErrEmptyDecimalStr      = errors.New("decimal string cannot be empty")
```
shouldn't these be moved to the `sdkerrors` pkg? so that they are imported as `sdkerrors.<...>` instead of `sdk.<...>`

alexanderbez replied:
no, they're typed errors for decimals. They're specific errors defined near the scope of the actual type(s).

fedekunze commented on store/rootmulti/store.go:
```diff
@@ -12,11 +12,11 @@ import (
 
 	"github.com/cosmos/cosmos-sdk/store/cachemulti"
 	"github.com/cosmos/cosmos-sdk/store/dbadapter"
-	"github.com/cosmos/cosmos-sdk/store/errors"
```
lol what was this?

alexanderbez replied:
lol it was to avoid import cycles 🤦‍♂ but its removed now!

AdityaSripal commented on types/result.go:
```diff
@@ -12,36 +12,27 @@ import (
-	Log string
-
+// GasInfo defines tx execution gas context.
+type GasInfo struct {
```
Include `BlockGasWanted uint64` here

alexanderbez replied:
See above; resolving.

alessio commented:
Looks good. Just a question (maybe even an out-of-scope one admittedly): for the sake of design cleanliness and simplicity, shouldn't we flatten the errors subpackage into types? What's the value of having a separate errors subpackage when it's likely that clients will use errors alongside the sdk common types?

fedekunze commented on baseapp/baseapp.go:
```diff
@@ -523,7 +524,8 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 	// only run the tx if there is block gas remaining
 	if mode == runTxModeDeliver && ctx.BlockGasMeter().IsOutOfGas() {
-		return sdk.ErrOutOfGas("no block gas left to run tx").Result()
+		gInfo = sdk.GasInfo{GasUsed: ctx.BlockGasMeter().GasConsumed()}
```
++ although that’s out of scope from this PR

AdityaSripal replied:
I don't think so. Since `GasInfo` is being introduced in this PR; i think it's a relevant requested change for this PR

alexanderbez replied:
This is a good observation. I believe this proposal would improve understanding of the semantics. I'll update accordingly.

alexanderbez replied:
Synced with @AdityaSripal and we've agreed that the suggestion doesn't make much sense without the ability to modfiy `ResponseCheckTx` and `ResponseDeliverTx`. Resolving this issue. @AdityaSripal will create an issue in Tendermint.

alexanderbez commented on baseapp/abci.go:
```diff
@@ -158,50 +159,94 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
+	gInfo, result, err := app.runTx(mode, req.Tx, tx)
+	if err != nil {
+		space, code, log = sdkerrors.ABCIInfo(err, false)
+	} else if result != nil {
```
The semantics we want to main is that if any error occurs, that `result` is always `nil`. There would be no reason to return a `*Result` as well because what context would it contain that would be needed? The error contains all relevant ABCI info.

alexanderbez commented on baseapp/baseapp.go:
```diff
@@ -568,32 +575,33 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
+	msgs := tx.GetMsgs()
 	if err := validateBasicTxMsgs(msgs); err != nil {
-		return err.Result()
+		gInfo = sdk.GasInfo{GasUsed: ctx.BlockGasMeter().GasConsumed()}
```
See above; resolving.

alexanderbez commented on baseapp/baseapp.go:
```diff
@@ -602,10 +610,8 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-			res.GasUsed = ctx.GasMeter().GasConsumed()
-			return res
+			gInfo = sdk.GasInfo{GasWanted: gasWanted, GasUsed: ctx.GasMeter().GasConsumed()}
+			return gInfo, nil, err
```
Yes, you're correct. This statement is redundant. I removed it.

alexanderbez commented on baseapp/baseapp.go:
```diff
@@ -615,83 +621,63 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 
-	return result
+	gInfo = sdk.GasInfo{GasWanted: gasWanted, GasUsed: runMsgCtx.GasMeter().GasConsumed()}
+	return gInfo, result, err
```
updated.

alexanderbez commented:
@alessio Yes, we could, but that would cause an import mess with `store/` so I'd opt to leave as-is for now.

@fedekunze and @AdityaSripal, I addressed all comments.

AdityaSripal commented on baseapp/baseapp.go:
```diff
@@ -610,7 +610,6 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (gInfo sdk.
 
 		if err != nil {
-			gInfo = sdk.GasInfo{GasWanted: gasWanted, GasUsed: ctx.GasMeter().GasConsumed()}
 			return gInfo, nil, err
```
nit: A simple `return` would suffice here correct?

alexanderbez replied:
you mean a naked return? Yes, but that drastically reduces comprehensibility.  Naked returns should only be used in very short/concise methods & functions.

ethanfrey commented on baseapp/abci.go:
```diff
@@ -158,50 +159,94 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
+func (app *BaseApp) CheckTx(req abci.RequestCheckTx) abci.ResponseCheckTx {
 	tx, err := app.txDecoder(req.Tx)
+	if err != nil {
+		space, code, log := sdkerrors.ABCIInfo(err, false)
```
Logic looks good, but I'm sure you will be doing this a few places.
I would recommend adding `CheckTxError(err error) abci.ResponseCheckTx` function (in types? in baseapp?) and then using it these places to make error handling less of a chore

alexanderbez replied:
Added functions to convert errors to the correct ABCI response types.

ethanfrey commented on baseapp/abci.go:
```diff
@@ -158,50 +159,94 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
-		GasWanted: int64(result.GasWanted), // TODO: Should type accept unsigned ints?
-		GasUsed:   int64(result.GasUsed),   // TODO: Should type accept unsigned ints?
-		Events:    result.Events.ToABCIEvents(),
+		GasWanted: int64(gInfo.GasWanted), // TODO: Should type accept unsigned ints?
```
Okay, so the purpose to combine these two is for the gas Wanted/Used?

Maybe update the function I proposed above to take `(err error, gas GasInfo)` and then it will work for the top case (0 gas) and the bottom case. And you can just short-circuit return in the err != nil phase, rather than this slightly-confusing flow control

alexanderbez replied:
Wasn't a fan of the control flow either. I'll tweak it some more.

ethanfrey commented on baseapp/abci.go:
```diff
@@ -158,50 +159,94 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
-		result = err.Result()
-	} else {
-		result = app.runTx(runTxModeDeliver, req.Tx, tx)
+		space, code, log := sdkerrors.ABCIInfo(err, false)
```
Same point on helper function, I guess `DeliverTxError` needs to be different than `CheckTxError`, but I still think they make this clearer

ethanfrey commented on baseapp/abci.go:
```diff
@@ -294,61 +338,59 @@ func (app *BaseApp) Query(req abci.RequestQuery) (res abci.ResponseQuery) {
 
-	msg := "unknown query path"
-	return sdk.ErrUnknownRequest(msg).QueryResult()
+	return sdkerrors.QueryResult(sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, "unknown query path"))
```
Okay, you already have a QueryResult helper. I like that.

ethanfrey commented on baseapp/baseapp.go:
```diff
@@ -535,19 +537,25 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-				log := fmt.Sprintf(
-					"out of gas in location: %v; gasWanted: %d, gasUsed: %d",
-					rType.Descriptor, gasWanted, ctx.GasMeter().GasConsumed(),
+				err = sdkerrors.Wrap(
```
If `sdk.ErrorOutOfGas` were an `sdkerrors` type, then we would not even need to cast it here. Just `err = rType`.

Both for consistency/aesthertics, but also one thing lost here is the stack trace. The original error that was thrown in panic would have the original stack trace. Maybe this is a minor point, but maybe interesting for development and likely minimal work to change.

alexanderbez replied:
Added a TODO to address in a subsequent PR.

ethanfrey commented on baseapp/baseapp.go:
```diff
@@ -535,19 +537,25 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 			default:
-				log := fmt.Sprintf("recovered: %v\nstack:\n%v", r, string(debug.Stack()))
-				result = sdk.ErrInternal(log).Result()
+				err = sdkerrors.Wrap(
```
:+1:

ethanfrey commented on baseapp/baseapp.go:
```diff
@@ -615,83 +620,62 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
-		data = append(data, msgResult.Data...)
-
-		msgEvents := msgResult.Events
+		data = append(data, result.Data...)
```
I figured that we should expect exactly one of `result` or `err` is non-nil. That is suggested from:
`type Handler func(ctx Context, msg Msg) (*Result, error)`.
If `Result` must always be non-nil, please make it a struct rather than a pointer. At the least explicitly check if `result` is non-nil before reading any values from it.

My personal preference if just to return an error on a fail. If we have multiple messages, logs/data/events of the non-failure ones can be recorded, but I don't see why we expect any info from the failed message. This did bring up some confusion to other devs on slack as well.

That said, if you think it is better this way, that is fine. Just make sure to check or enforce the result is not nil

alexanderbez replied:
Yes, great catch! We need to make `runMsgs` an exception -- it being a private method allows to developers not to have to worry about this.

Since we _always_ want to return a `Result` regardless of message execution success or failure, I agree we should change the signature to `sdk.Result, error` and always return a `Result`. This allows semantics to remain unchanged as `runTx` will not commit state upon `runMsgs` error.

alexanderbez replied:
On further thought, this really makes semantics hard to understand. If I as the developer and enforcer of these semantics find it difficult, then surely it'll be difficult for upstream users.

Thus, I believe we should go with the simple and intuitive approach: always return `*Result` or `error`, but never both. This means that if any single message fails, we'll only retain data for that
failed message in the error -- losing any info on previous successful messages and I think this is OK. This is what you were trying to achieve the entire time @ethanfrey.

ethanfrey commented on baseapp/baseapp.go:
```diff
@@ -615,83 +620,62 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 	}
 
-	return result
+	return &sdk.Result{
```
This is odd... so if one msg returns an error, you update the logs and break processing (line 670), but still return err=nil to runTx. Make sure to return an error in that case. Otherwise, this will commit the prior messages (imagine msg 1 succeeded but msg 2 failed, this must return an error to cause the state changes from msg 1 to be reverted)

alexanderbez replied:
Yes, this is a bug. See response in https://github.com/cosmos/cosmos-sdk/pull/5421#discussion_r360994196. Thanks for the review and catching this!

alexanderbez commented:
@ethanfrey @AdityaSripal @fedekunze just to reiterate from my response https://github.com/cosmos/cosmos-sdk/pull/5421#discussion_r361005897, in order to drastically reduce semantic confusion and flow, the following should always hold true:

- A `*Result` or `error` is returned from `runMsgs` or `runTx` but **never both**.
- State transitions only occur if **no error** is returned on `runMsgs`.
- Even if a single message fails out of `n`, a single error is returned (context on successful messages are lost).
  - This is the only break to the current semantics but it makes things much easier to read and handle.

As a result, `DeliverTx` and `CheckTx` responses will return ABCI info from the error or the result, but never both. An error if gas issues occurred or message execution failed. Otherwise, a `*Result` is guaranteed.

marbar3778 commented:
should there be a followup PR from docs to update docs on this? I think in some places we are copying in code which uses the old way

alexanderbez commented:
I updated some docs with respect to `Handler` and `Querier`, but I'm certain there are remnants of `sdk.Error` in docs somewhere. So we should definitely do a sweep through!
